app/src/main/java/ar/scacchipa/twittercloneapp/viewmodel/LoginApiViewModel.kt

```diff
@@ -0,0 +1,49 @@
+package ar.scacchipa.twittercloneapp.viewmodel
+
+import android.net.Uri
+import androidx.core.net.toUri
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import ar.scacchipa.twittercloneapp.datasource.UserAccessToken
+import ar.scacchipa.twittercloneapp.datasource.provideAuthSourceDateApi
+import ar.scacchipa.twittercloneapp.datasource.provideRetrofit
+import ar.scacchipa.twittercloneapp.domain.AuthorizationUseCase
+import ar.scacchipa.twittercloneapp.repository.AuthorizationRepository
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+class LoginApiViewModel (
+    private val userCase: AuthorizationUseCase =
+        AuthorizationUseCase(AuthorizationRepository(provideAuthSourceDateApi(provideRetrofit())))
+): ViewModel() {
+    var userAccessToken = MutableLiveData<UserAccessToken>()
+        private set
+
+    fun controlRequest(uri: Uri, clientId: String, redirectUri: String) {
+        if (uri.host == redirectUri.toUri().host) {
+            uri.getQueryParameter("code")?.let { code ->
+                this.generateUserAccessToken(code, clientId, redirectUri)
+            }
+        }
+    }
+
+    private fun generateUserAccessToken(transitoryCode: String, clientId: String, redirectUri: String) {
+        viewModelScope.launch {
+            userAccessToken.value = withContext(Dispatchers.IO) {
```

Comment:
Remember try to not hardcode Dispatchers, then it will be a nightmare when unit testing. Also I would not change dispatchers yet, I would delay that switching in the data layer when it is needed.

Reply:
"withContext(Dispatchers.IO)" was move to AuthorizationUseCase class.

Reply:
The withContext(Dispatchers.IO) was shifted to a repository class